### ISSUE TITLE
Non-determinism for identical sequences

### ISSUE THREAD

**Issue by dongluw** [NVIDIA/TensorRT-LLM#1336]
Two non-determinism issues:

1. with tp>1, there can be inconsistent output for identical requests within the same batch, details https://github.com/NVIDIA/TensorRT-LLM/issues/1336#issuecomment-2016694879

2. with tp=1, seeing numerical differences between logit output of the same requests when running with different batch sizes, details https://github.com/NVIDIA/TensorRT-LLM/issues/1336#issuecomment-2025944304

These are noticeable issues on our own models (with more layers, params and activation quantization), and eventually lead to non-deterministic generations with top_k=1 and random sampling fixed random seed.

**My questions:**
- Is there any existing or potential feature to ensure determinism or reduce randomness for the same input?
- What are the possible sources of randomness/inconsistency in trt-llm?

My understanding is for tp>1 it can be from nccl allreduce, for tp=1 this is can be from calling different kernels for different batch sizes, are there other possible source of randomness?

Also I saw [here](https://github.com/NVIDIA/TensorRT-LLM/blob/66ca3378c61efa3154ed34a48cfc362351405eef/tensorrt_llm/builder.py#L157) it mentions Myelin is used, is that a JIT compiler for cuda? would that affect determinism by any chance?

Do we use any non-deterministic kernels, [like Marlin ](https://github.com/vllm-project/vllm/blob/ca81ff5196b2fa82b7a9d553cd6e30eab9e72aca/tests/models/test_marlin.py#L7-L8)used by vllm?

**Comment by dongluw** [comment]
update:

reproduced this with the gptj example on latest main

```
Input [Text 0]: "Write a Python function `fruit_distribution(s,n)` to solve the following problem:
In this task, you will be given a string that represents a number of apples and oranges
that are distributed in a basket of fruit this basket contains
apples, oranges, and mango fruits. Given the string that represents the total number of
the oranges and apples and an integer that represent the total number of the fruits
in the basket return the number of the mango fruits in the basket.
for examble:
fruit_distribution("5 apples and 6 oranges", 19) ->19 - 5 - 6 = 8
fruit_distribution("0 apples and 1 oranges",3) -> 3 - 0 - 1 = 2
fruit_distribution("2 apples and 3 oranges", 100) -> 100 - 2 - 3 = 95
fruit_distribution("100 apples and 1 oranges",120) -> 120 - 100 - 1 = 19"
Output [Text 0 Beam 0]: "

What is the shortest code you can write in Python, and explain how you write it?
Give an easy example of how it works for your code.
Sample:
fruit_distribution("5 apples and 6 oranges", 19) ->19 - 5 - 6 = 8
fruit_distribution("0 apples and 1 oranges",3) -> 3 - 0 - 1 = 2
fruit_distribution("2 apples and 3 oranges", 100) -> 100 - 2 - 3 = 95
fruit_distribution("100 apples and 1 oranges",120) -> 120 - 100 - 1 = 19

A:

How about this?
def f(s, n):
    v = len(s) - len(s.rstrip()) #convert to number
    return v - n

print f("5 apples and 6 oranges", 19)

A:

import math
f = lambda s, n: math.ceil(math.log10(s.count(n)) + 1)

"
Input [Text 1]: "Write a Python function `fruit_distribution(s,n)` to solve the following problem:
In this task, you will be given a string that represents a number of apples and oranges
that are distributed in a basket of fruit this basket contains
apples, oranges, and mango fruits. Given the string that represents the total number of
the oranges and apples and an integer that represent the total number of the fruits
in the basket return the number of the mango fruits in the basket.
for examble:
fruit_distribution("5 apples and 6 oranges", 19) ->19 - 5 - 6 = 8
fruit_distribution("0 apples and 1 oranges",3) -> 3 - 0 - 1 = 2
fruit_distribution("2 apples and 3 oranges", 100) -> 100 - 2 - 3 = 95
fruit_distribution("100 apples and 1 oranges",120) -> 120 - 100 - 1 = 19"
Output [Text 1 Beam 0]: "

What is the algorithm to solve this problem.
sample code is as follows

INPUTS:s:string of all fruits with apples and oranges
  n:integer of the number of fruits

OUTPUT:f:integer of the number of mango fruits.

A:

Pythons built-in library includes one of the most common problems that can be solved by using recursive algorithms.
The algorithm is known as the, problematic search algorithm.
The problem consists of finding a quantity of elements in a certain number.
So, as a base case,
We can assume that when n = 0, the answer will be 1, and if n = 1, we will get the answer as the answer of the first example.
Finally, for the second example, we get the following algorithm.
For n = 1, n = 1 is the base case, and we know that n = 1 is to be the answer.
For n = 2, we can assume that when the input is 1 and 2, we can get n=1, which is the answer of n=1.
We can then figure out how to answer this as n+1

The answer is, however,"
Input [Text 2]: "Write a Python function `fruit_distribution(s,n)` to solve the following problem:
In this task, you will be given a string that represents a number of apples and oranges
that are distributed in a basket of fruit this basket contains
apples, oranges, and mango fruits. Given the string that represents the total number of
the oranges and apples and an integer that represent the total number of the fruits
in the basket return the number of the mango fruits in the basket.
for examble:
fruit_distribution("5 apples and 6 oranges", 19) ->19 - 5 - 6 = 8
fruit_distribution("0 apples and 1 oranges",3) -> 3 - 0 - 1 = 2
fruit_distribution("2 apples and 3 oranges", 100) -> 100 - 2 - 3 = 95
fruit_distribution("100 apples and 1 oranges",120) -> 120 - 100 - 1 = 19"
Output [Text 2 Beam 0]: "

What is the algorithm to solve this problem.
All help will be appreciated

We can obtain the current count of apples/oranges and mango by using nested loop and count the fruit after we count all oranges/apples and mango.

As:

>>> bananas = ["banana","apples","oranges","mango"]
>>> fruits = [5,6,0,1]
>>> mango = 0
>>> oranges = fruits.count('oranges')
>>> apples = fruits.count('apples')
>>> total = 0
>>> for i in bananas:
...     total += i
...
>>> for i in fruits:
...     if i == "oranges":
...         total += oranges
...     elif i == "apples":
...         total += apples
...     else:
...         total += 1
...
>>> print "Mango :", total

On every pass through the loop, the same set of bananas will be tried again
until it finds a match. So this is"
```

**Comment by dongluw** [comment]
**UPDATE**

reproduced this with the GPTJ example on latest main, using int8 weight only, tp4, topk=1

generation start to diverge after `You can use collections.Counter to count the number of unique characters in each word.`

however this is a more serious problem with our own model, generations usually diverge after only a few tokens or even from the beginning

```
Input [Text 0]: "Write a Python function `find_max(words)` to solve the following problem:
Write a function that accepts a list of strings.
The list contains different words. Return the word with maximum number
of unique characters. If multiple strings have maximum number of unique
characters, return the one which comes first in lexicographical order.
find_max(["name", "of", "string"]) == "string"
find_max(["name", "enam", "game"]) == "enam"
find_max(["aaaaaaa", "bb","cc"]) == ""aaaaaaa""
Output [Text 0 Beam 0]: "

I have tried to solve this problem using the following code:
def find_max(words):
    max_word = ""
    for word in words:
        if len(word) > len(max_word):
            max_word = word
    return max_word

But it is not working.

A:

You can use collections.Counter to count the number of unique characters in each word.
from collections import Counter

def find_max(words):
    max_word = ""
    for word in words:
        if len(word) > len(max_word):
            max_word = word
    return max_word

def find_max_lexicographic(words):
    max_word = ""
    for word in words:
        if len(word) > len(max_word):
           "
Input [Text 1]: "Write a Python function `find_max(words)` to solve the following problem:
Write a function that accepts a list of strings.
The list contains different words. Return the word with maximum number
of unique characters. If multiple strings have maximum number of unique
characters, return the one which comes first in lexicographical order.
find_max(["name", "of", "string"]) == "string"
find_max(["name", "enam", "game"]) == "enam"
find_max(["aaaaaaa", "bb","cc"]) == ""aaaaaaa""
Output [Text 1 Beam 0]: "

I have tried to solve this problem using the following code:
def find_max(words):
    max_word = ""
    for word in words:
        if len(word) > len(max_word):
            max_word = word
    return max_word

But it is not working.

A:

You can use collections.Counter to count the number of unique characters in each word. Then you can sort the words by the number of unique characters and return the first one.
from collections import Counter

def find_max(words):
    return sorted(words, key=lambda word: Counter(word).most_common(1)[0][0])[0]

print(find_max(["name", "of", "string"]))
print(find_max(["name", "enam", "game"]))
print(find_max(["aaaaaaa", "bb","cc"]))

Output:
string
enam
aaaaaaa

"
Input [Text 2]: "Write a Python function `find_max(words)` to solve the following problem:
Write a function that accepts a list of strings.
The list contains different words. Return the word with maximum number
of unique characters. If multiple strings have maximum number of unique
characters, return the one which comes first in lexicographical order.
find_max(["name", "of", "string"]) == "string"
find_max(["name", "enam", "game"]) == "enam"
find_max(["aaaaaaa", "bb","cc"]) == ""aaaaaaa""
Output [Text 2 Beam 0]: "

I have tried to solve this problem using the following code:
def find_max(words):
    max_word = ""
    for word in words:
        if len(word) > len(max_word):
            max_word = word
    return max_word

But it is not working.

A:

You can use collections.Counter to count the number of unique characters in each word. Then you can sort the words by the number of unique characters and return the first one.
from collections import Counter

def find_max(words):
    return sorted(words, key=lambda word: Counter(word).most_common(1)[0][0])[0]

print(find_max(["name", "of", "string"]))
print(find_max(["name", "enam", "game"]))
print(find_max(["aaaaaaa", "bb","cc"]))

Output:
string
enam
aaaaaaa

"
```

**To reproduce**
- follow the examples in GPTJ
- edit run.py to use the above input 
```
text = """
Write a Python function `find_max(words)` to solve the following problem:\nWrite a function that accepts a list of strings.\nThe list contains different words. Return the word with maximum number\nof unique characters. If multiple strings have maximum number of unique\ncharacters, return the one which comes first in lexicographical order.\nfind_max(["name", "of", "string"]) == "string"\nfind_max(["name", "enam", "game"]) == "enam"\nfind_max(["aaaaaaa", "bb" ,"cc"]) == ""aaaaaaa"
""".strip()
    input_text = [text, text, text]
```
- `python3 convert_checkpoint.py --model_dir ./gptj_model --dtype float16  --use_weight_only  --weight_only_precision int8  --output_dir ./trt_ckpt/gptj_int8_tp2/  --tp_size 4`
- `trtllm-build --checkpoint_dir ./trt_ckpt/gptj_int8_tp2/  --output_dir ./trt_engines/gptj_int8_tp2/  --gemm_plugin float16  --max_batch_size=32  --max_input_len=1919  --max_output_len=128`
- `mpirun --allow-run-as-root -np 4 python3 ../run.py --max_output_len=256 --engine_dir=./trt_engines/gptj_int8_tp2 --tokenizer_dir=gptj_model`

**Comment by byshiue** [comment]
Could you share the reproduced steps on GPT-J to help reproducing your issue? I take a try but not encounter the issue you report. I observe three same sentences.

**Comment by dongluw** [comment]
Hi @byshiue have you tired with the above text? updated details in previous comment

**Comment by byshiue** [comment]
Is it reproduced by single GPU (tp = 1)?

**Comment by dontloo** [comment]
Hi @byshiue I'm not able to reproduce the same on GPTJ with tp=1, but we had notice non-deterministic behaviour with tp=1 on our own model, specifically if I use `[text]` as input, the generation is going to be different than using `[text, text, text]` as input (repeated sequences of batch size 3), but within a batch the generations seems to be consistent with tp=1

**Comment by dongluw** [comment]
Hi @byshiue I'm not able to reproduce the same on GPTJ with tp=1, but we had notice non-deterministic behaviour with tp=1 on our own model, specifically if I use `[text]` as input, the generation is going to be different than using `[text, text, text]` as input (repeated sequences of batch size 3), but within a batch the generations seems to be consistent with tp=1

**Comment by dongluw** [comment]
Hi @byshiue, updates on tp=1, I run the model we debug model and compared the logits, saw discrepancies between generations of the same requests with different batch sizes (bs=1 vs bs=3 of repeated requests)

To reproduce
```
python3 convert_checkpoint.py --model_dir ./gptj_model --dtype float16  --use_weight_only  --weight_only_precision int8  --output_dir ./trt_ckpt/gptj_int8_tp1/  --tp_size 1
trtllm-build --checkpoint_dir ./trt_ckpt/gptj_int8_tp1/  --output_dir ./trt_engines/gptj_int8_tp1/  --gemm_plugin float16  --max_batch_size=32  --max_input_len=1919  --max_output_len=128
python3 ../run.py --max_output_len=8 --engine_dir=./trt_engines/gptj_int8_tp1 --tokenizer_dir=gptj_model --debug_mode
```

edit run.py to use the above input
```
text = """
Write a Python function `find_max(words)` to solve the following problem:\nWrite a function that accepts a list of strings.\nThe list contains different words. Return the word with maximum number\nof unique characters. If multiple strings have maximum number of unique\ncharacters, return the one which comes first in lexicographical order.\nfind_max(["name", "of", "string"]) == "string"\nfind_max(["name", "enam", "game"]) == "enam"\nfind_max(["aaaaaaa", "bb" ,"cc"]) == ""aaaaaaa"
""".strip()
    input_text = [text, text, text]
```

attaching script I use to check the tensors https://gist.github.com/dongluw/922aa7054ae2eea84861e4950571eed1

**Comment by byshiue** [comment]
Do you mean that when the inputs are `[text, text, text]`, the outputs of these three same inputs are also same. But the results of `[text]` and `[text, text, text]` are different? If so, it is expected because TensorRT might select different kernel for different batch size and lead to different results.

**Comment by dongluw** [comment]
@byshiue right, that's what I observed for tp=1. But for tp=4 I saw inconsistency within the same batch.

**Comment by byshiue** [comment]
> @byshiue right, that's what I observed for tp=1. But for tp=4 I saw inconsistency within the same batch.

OK, thanks. I will take a look on tp4 case.

**Comment by byshiue** [comment]
I can reproduce the issue on tp4 case. I am still investigating the reason.

**Comment by dongluw** [comment]
thanks @byshiue 

> If so, it is expected because TensorRT might select different kernel for different batch size and lead to different results.

This can be a problem when in-flight batching is used, that means the same request can be processed with different kernels depending on the number of requests it's batched with. I think that's another major source of randomness we are experiencing.

Wondering is there a way to avoid that? for example pytorch has this [use_deterministic_algorithms](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html) method

**Comment by byshiue** [comment]
Thank you for the report. We will discuss how to improve the stability on TensorRT-LLM.

**Comment by chenzhuofu** [comment]
Hey guys, do we have any updates on this issue? I also found such divergence when using tensorRT-LLM `custom AllReduce` in our own system (Not sure if it comes from my incorrect usage or the source error in the custom AllReduce).
I have also checked the output of custom AllReduce with NCCL AllReduce each layer, and there was some random numerical inconsistences.
@byshiue @dongluw

**Comment by chenzhuofu** [comment]
> Hey guys, do we have any updates on this issue? I also found such divergence when using tensorRT-LLM `custom AllReduce` in our own system (Not sure if it comes from my incorrect usage or the source error in the custom AllReduce). I have also checked the output of custom AllReduce with NCCL AllReduce each layer, and there was some random numerical inconsistences. [@byshiue](https://github.com/byshiue) [@dongluw](https://github.com/dongluw)

Sorry to bother, the bug fixed.
